banner_ics.php

```diff
@@ -63,20 +64,47 @@ public function process_attachment(&$content, &$message, &$a)
             // Make sure we have events
             if (!$ical->hasEvents()) return;
 
+            // Crete timezone objects for calendar timezone and RoundCube UI timezone
+            try {
+                $ui_tz = new DateTimeZone($rcmail->config->get('timezone'));
+            } catch (Exception $e) {
+                $ui_tz = new DateTimeZone(DateTimeZone::UTC);
+            }
+            try {
+                $ical_tz = new DateTimeZone($ical->calendarTimeZone(true));
+            } catch (Exception $e) {
+                // Use UI timezone if the calendar have no timezone specified
+                $ical_tz = $ui_tz;
+            }
+
             // Get first event
             foreach ($ical->events() as &$event) {
-                $dtstart = $event->dtstart_array[2];
-                $dtend = $event->dtend_array[2];
-                $dtstr = $rcmail->format_date($dtstart, $combined_format) . ' - ';
-
-                // Dont double date if same
-                $df = 'Y-m-d';
-                if (date($df, $dtstart) === date($df, $dtend)) {
-                    $dtstr .= $rcmail->format_date($dtend, $time_format);
+                $is_all_day = isset($event->dtstart_array[0]['VALUE']) && $event->dtstart_array[0]['VALUE'] === 'DATE';
+
+                if ($is_all_day) {
+                    // All day events should use UI timezone to avoid turning dates
+                    $dtstart = new DateTime($event->dtstart, $ui_tz);
+                    $dtend = new DateTime($event->dtend, $ui_tz);
+                    // All day events ends at next day midnight, we should fix the day
+                    $dtend->modify('-1 day');
```

Comment:
I don't think this is right. From RFC 5545 (https://tools.ietf.org/html/rfc5545#page-54)

      The "VEVENT" is also the calendar component used to specify an
      anniversary or daily reminder within a calendar.  These events
      have a DATE value type for the "DTSTART" property instead of the
      default value type of DATE-TIME.  If such a "VEVENT" has a "DTEND"
      property, it MUST be specified as a DATE value also.  The
      anniversary type of "VEVENT" can span more than one date (i.e.,
      "DTEND" property value is set to a calendar date after the
      "DTSTART" property value).  If such a "VEVENT" has a "DURATION"
      property, it MUST be specified as a "dur-day" or "dur-week" value.

From what I understand, `DTEND` would be absent in a single day event, so we need to check for that instead. Also see the examples on Pg. 55